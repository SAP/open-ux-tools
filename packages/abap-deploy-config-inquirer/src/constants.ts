export const ABAP_PACKAGE_SEARCH_MAX_RESULTS = 50;
export const DEFAULT_PACKAGE_ABAP = '$tmp';
