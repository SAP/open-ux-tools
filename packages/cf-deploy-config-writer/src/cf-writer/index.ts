export { generateAppConfig } from './app-config';
export { generateRootConfig } from './base-config';
