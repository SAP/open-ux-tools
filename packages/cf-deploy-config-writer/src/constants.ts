import { UI5_DEFAULT } from '@sap-ux/ui5-config';
import { t } from './i18n';

export const WelcomeFile = 'welcomeFile';
export const XSAppFile = 'xs-app.json';
export const XSSecurityFile = 'xs-security.json';
export const MTABuildResult = 'build-result';
export const MTABuildParams = 'build-parameters';
export const MTAFileExtension = 'mta-ext.mtaext';
export const DefaultServiceURL = '${default-url}';
export const ManagedXSUAA = 'managed:xsuaa';
export const HTML5RepoHost = 'html5-apps-repo:app-host';
export const SRV_API = 'srv-api';
export const DefaultMTADestination = 'fiori-default-srv-api';
export const EmptyDestination = 'NONE';
export const ResourceMTADestination = '%s-srv-api';
export const MTAYamlFile = 'mta.yaml';
export const MTADescription = 'Generated by Fiori Tools';
export const RouterModule = 'router';
export const CloudFoundry = 'cf';
export const deployMode = 'deploy_mode';
export const enableParallelDeployments = 'enable-parallel-deployments';
export const CDSExecutable = 'cds';
export const CDSPackage = '@sap/cds-dk';
export const MTAExecutable = 'mta';
export const MTAPackage = 'mta';
export const MTAPackageVersion = '^1.2.27';
export const MTAVersion = '0.0.1';
export const RimrafVersion = '^5.0.5';
export const Rimraf = 'rimraf';
export const MbtPackage = 'mbt';
export const MbtPackageVersion = '^1.2.29';
export const UI5BuilderWebIdePackage = '@sap/ui5-builder-webide-extension';
export const UI5BuilderWebIdePackageVersion = '^1.1.9';
export const UI5TaskZipperPackage = 'ui5-task-zipper';
export const UI5TaskZipperPackageVersion = '^3.1.3';
export const UI5Package = '@ui5/cli';
export const UI5PackageVersion = '^3.9.2';
export const CDSAddMtaParams = ['add', 'mta'];
export const CDSXSUAAService = 'xsuaa';
export const CDSHTML5RepoService = 'html5-repo';
export const CDSConnectivityService = 'connectivity';
export const CDSDestinationService = 'destination';
export const MTAAPIDestination = {
    Name: ResourceMTADestination,
    Type: 'HTTP',
    URL: `~{srv-api/srv-url}`,
    ProxyType: 'Internet',
    Authentication: 'NoAuthentication',
    'HTML5.DynamicDestination': true,
    'HTML5.ForwardAuthToken': true
};
export const UI5Destination = {
    Name: 'ui5',
    Type: 'HTTP',
    URL: UI5_DEFAULT.SAPUI5_CDN,
    ProxyType: 'Internet',
    Authentication: 'NoAuthentication'
};
export const UI5ResourceDestination = {
    'init_data': {
        instance: {
            'existing_destinations_policy': 'update',
            destinations: [
                {
                    Name: 'ui5',
                    Type: 'HTTP',
                    URL: UI5_DEFAULT.SAPUI5_CDN,
                    ProxyType: 'Internet',
                    Authentication: 'NoAuthentication'
                }
            ]
        }
    }
};

export const UI5StandaloneModuleDestination = {
    group: 'destinations',
    properties: {
        forwardAuthToken: false,
        name: 'ui5',
        url: UI5_DEFAULT.SAPUI5_CDN
    }
};

export const DestinationServiceConfig = {
    config: {
        'HTML5Runtime_enabled': true,
        version: '1.0.0',
        ...UI5ResourceDestination
    }
};

export const ServiceAPIRequires = {
    name: SRV_API,
    properties: {
        'srv-url': DefaultServiceURL
    }
};
export const HTMLAppBuildParams = {
    builder: 'custom',
    'build-result': 'dist',
    commands: ['npm install', 'npm run build:cf'],
    'supported-platforms': []
};
export const UI5DeployBuildScript =
    'ui5 build preload --clean-dest --config ui5-deploy.yaml --include-task=generateCachebusterInfo';
export const MTABuildScript = 'rimraf resources mta_archives && mbt build';
export const appDeployMTAScript = (args: string[]): string => {
    const mtaArgs = args.length > 0 ? ` ${args.join(' ')}` : '';
    return `fiori cfDeploy${mtaArgs}`;
};
export const rootDeployMTAScript = (args: string[]): string => {
    const mtaArgs = args.length > 0 ? `${args.join(' ')} ` : '';
    return `cf deploy mta_archives/archive.mtar ${mtaArgs}--retries 1`;
};
export const undeployMTAScript = (mtaId: string): string =>
    `cf undeploy ${mtaId} --delete-services --delete-service-keys --delete-service-brokers`;
const cannotFindBinary = (bin: string, pkg: string): string => t('error.cannotFindBinary', { bin, pkg });
export const CDSBinNotFound = cannotFindBinary(CDSExecutable, CDSPackage);
export const MTABinNotFound = cannotFindBinary(MTAExecutable, MTAPackage);
