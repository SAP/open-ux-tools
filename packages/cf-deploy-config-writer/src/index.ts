export * from './mta-config';
export * from './cf-writer';
export { CFBaseConfig, CFAppConfig, RouterModuleType, ApiHubConfig } from './types';
