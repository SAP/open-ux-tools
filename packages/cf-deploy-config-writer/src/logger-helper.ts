import { ToolsLogger, type Logger } from '@sap-ux/logger';

/**
 * Static logger prevents passing of logger references through all functions, as this is a cross-cutting concern.
 */
export default class LoggerHelper {
    private static _logger: Logger = new ToolsLogger({ logPrefix: '@sap-ux/cf-deploy-config-writer' });

    /**
     * Get the logger.
     *
     * @returns the logger
     */
    public static get logger(): Logger {
        return LoggerHelper._logger;
    }

    /**
     * Set the logger.
     *
     * @param value the logger to set
     */
    public static set logger(value: Logger) {
        LoggerHelper._logger = value;
    }
}
