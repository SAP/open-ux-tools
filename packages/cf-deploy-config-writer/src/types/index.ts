import type { Destination, Authentication } from '@sap-ux/btp-utils';

export type ResourceType =
    | 'xsuaa'
    | 'destination'
    | 'portal'
    | 'connectivity'
    | 'managed:xsuaa'
    | 'html5-apps-repo:app-host'
    | 'html5-apps-repo:app-runtime';
export type ModuleType =
    | 'hdb'
    | 'nodejs'
    | 'java'
    | 'approuter.nodejs'
    | 'com.sap.application.content'
    | 'com.sap.application.content:destination'
    | 'com.sap.application.content:resource'
    | 'html5'
    | 'com.sap.portal.content';
export enum CloudFoundryServiceType {
    Existing = 'org.cloudfoundry.existing-service',
    Managed = 'org.cloudfoundry.managed-service'
}
export type MTADestinationType = Destination & {
    ServiceInstanceName: string;
    ServiceKeyName: string;
    'sap.cloud.service': string;
};
export enum RouterModuleType {
    Standard = 'standard',
    Managed = 'managed'
}
export interface MTABaseConfig {
    mtaId: string;
    mtaPath: string;
    mtaDescription?: string;
    mtaVersion?: string;
}
export interface CFBaseConfig extends MTABaseConfig {
    routerType: RouterModuleType;
    addConnectivityService?: boolean;
    addDestinationService?: boolean;
    abapServiceProvider?: {
        abapServiceName?: string;
        abapService?: string;
    };
}
export interface CFAppConfig {
    appPath: string;
    addManagedRouter?: boolean;
    destination?: string;
    apiHubConfig?: ApiHubConfig;
    addMTADestination?: boolean;
    serviceHost?: string; // Data service host
    lcapMode?: boolean;
}
export interface CFConfig extends CFAppConfig, CFBaseConfig {
    appId: string;
    rootPath: string;
    serviceBase: string;
    capRoot?: string;
    isCap?: boolean;
    servicePath?: string;
    firstServicePathSegment?: string;
    isFullUrlDest?: boolean;
    destinationAuthType?: Authentication;
    cloudServiceName?: string;
    isMtaRoot?: boolean;
}
export const enum ApiHubType {
    apiHub = 'API_HUB',
    apiHubEnterprise = 'API_HUB_ENTERPRISE'
}
export interface ApiHubConfig {
    apiHubKey: string;
    apiHubType: ApiHubType;
}
export interface XSAppRoute {
    source?: string;
    target?: string;
    destination?: string;
    csrfProtection?: boolean;
    scope?: string;
    service?: string;
    endpoint?: string;
    authenticationType?: string;
    dependency?: string;
}
export type XSAppRouteProperties = keyof XSAppRoute;
export interface XSAppDocument {
    authenticationMethod?: string;
    routes?: XSAppRoute[];
    welcomeFile?: string;
}
export interface HTML5App {
    path: string;
    'build-parameters': {
        builder: string;
        'build-result': string;
        commands: string[];
        'supported-platforms': string[];
    };
    name: string;
    type: string;
}
