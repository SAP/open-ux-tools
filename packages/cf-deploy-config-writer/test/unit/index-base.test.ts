import { join } from 'path';
import fsExtra from 'fs-extra';
import hasbin from 'hasbin';
import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';
import { NullTransport, ToolsLogger } from '@sap-ux/logger';
import { type CFBaseConfig, generateBaseConfig } from '../../src';
import { RouterModuleType } from '../../src/types';
import type { Editor } from 'mem-fs-editor';
import { MTABinNotFound } from '../../src/constants';
import { apiGetInstanceCredentials } from '@sap/cf-tools';

jest.mock('@sap-ux/btp-utils', () => ({
    ...jest.requireActual('@sap-ux/btp-utils'),
    isAppStudio: jest.fn(),
    listDestinations: jest.fn()
}));

jest.mock('@sap/cf-tools');
const apiGetInstanceCredentialsMock = apiGetInstanceCredentials as jest.Mock;

describe('CF Writer', () => {
    let unitTestFs: Editor;
    const logger = new ToolsLogger({
        transports: [new NullTransport()]
    });
    const outputDir = join(__dirname, '../test-output');

    beforeEach(() => {
        jest.resetAllMocks();
        jest.restoreAllMocks();
        unitTestFs = create(createStorage());
    });

    beforeAll(() => {
        jest.clearAllMocks();
        jest.spyOn(hasbin, 'sync').mockReturnValue(true);
        fsExtra.removeSync(outputDir);
    });

    afterAll(() => {
        jest.resetAllMocks();
    });

    describe('Generate Base Config - Standalone', () => {
        test('Generate deployment configs - standalone with connectivity service', async () => {
            const debugSpy = jest.spyOn(logger, 'debug');
            const mtaId = 'standalonewithconnectivityservice';
            const mtaPath = join(outputDir, mtaId);
            fsExtra.mkdirSync(outputDir, { recursive: true });
            fsExtra.mkdirSync(mtaPath);
            await generateBaseConfig(
                {
                    mtaPath,
                    mtaId,
                    routerType: RouterModuleType.Standard,
                    addConnectivityService: true
                },
                unitTestFs,
                logger
            );
            expect(debugSpy).toBeCalledTimes(1);
            expect(unitTestFs.dump(mtaPath)).toMatchSnapshot();
            // Since mta.yaml is not in memfs, read from disk
            expect(unitTestFs.read(join(mtaPath, 'mta.yaml'))).toMatchSnapshot();
        });

        test('Generate deployment configs - standalone with ABAP service provider', async () => {
            apiGetInstanceCredentialsMock.mockResolvedValue({
                credentials: {
                    endpoints: { TestEndPoint: '' },
                    'sap.cloud.service': 'TestService'
                }
            });
            const mtaId = 'standalonewithabapserviceprovider';
            const mtaPath = join(outputDir, mtaId);
            fsExtra.mkdirSync(outputDir, { recursive: true });
            fsExtra.mkdirSync(mtaPath);
            await generateBaseConfig(
                {
                    mtaPath,
                    mtaId,
                    routerType: RouterModuleType.Standard,
                    abapServiceProvider: {
                        abapService: 'abap-haas',
                        abapServiceName: 'Y11_00.0035'
                    }
                },
                unitTestFs,
                logger
            );
            expect(unitTestFs.dump(mtaPath)).toMatchSnapshot();
            // Since mta.yaml is not in memfs, read from disk
            expect(unitTestFs.read(join(mtaPath, 'mta.yaml'))).toMatchSnapshot();
        });
    });

    describe('Generate Base Config - Managed', () => {
        test('Generate deployment configs - managed', async () => {
            const debugSpy = jest.spyOn(logger, 'debug');
            const mtaId = 'managed';
            const mtaPath = join(outputDir, mtaId);
            fsExtra.mkdirSync(outputDir, { recursive: true });
            fsExtra.mkdirSync(mtaPath);
            await generateBaseConfig(
                {
                    mtaPath,
                    mtaId,
                    mtaDescription: 'MyManagedDescription',
                    routerType: RouterModuleType.Managed
                },
                unitTestFs,
                logger
            );
            expect(debugSpy).toBeCalledTimes(1);
            expect(unitTestFs.dump(mtaPath)).toMatchSnapshot();
            // Since mta.yaml is not in memfs, read from disk
            expect(unitTestFs.read(join(mtaPath, 'mta.yaml'))).toMatchSnapshot();
        });
    });

    describe('Generate Base Config - Validation', () => {
        test('Generate invalid deployment configs', async () => {
            const mtaId = 'invalidconfigs02';
            const mtaPath = join(outputDir, mtaId);
            const config = {
                abapServiceProvider: {
                    abapService: '~abapService',
                    abapServiceName: '~abapService'
                },
                mtaPath,
                mtaId,
                mtaDescription: 'MyManagedDescription',
                routerType: RouterModuleType.Managed
            } as Partial<CFBaseConfig>;
            jest.spyOn(unitTestFs, 'exists').mockReturnValueOnce(true);
            await expect(generateBaseConfig(config as CFBaseConfig, unitTestFs)).rejects.toThrowError(
                'A folder with same name already exists in the target directory'
            );
            delete config.abapServiceProvider?.abapService;
            await expect(generateBaseConfig(config as CFBaseConfig)).rejects.toThrowError(
                'Missing ABAP service details for direct service binding'
            );
            await expect(generateBaseConfig({ ...config, mtaId: '~sample' } as CFBaseConfig)).rejects.toThrowError(
                'The MTA ID can only contain letters, numbers, dashes, periods and underscores (but no spaces)'
            );
            delete config.routerType;
            await expect(generateBaseConfig(config as CFBaseConfig)).rejects.toThrowError(
                'Missing required parameters, MTA path, MTA ID or router type'
            );
            jest.spyOn(hasbin, 'sync').mockReturnValue(false);
            await expect(generateBaseConfig(config as CFBaseConfig)).rejects.toThrowError(MTABinNotFound);
        });
    });
});
