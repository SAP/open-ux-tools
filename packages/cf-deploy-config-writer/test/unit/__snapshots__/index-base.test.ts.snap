// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CF Writer Base Generate Base Config - App Frontend Generate deployment configs - app frontend 1`] = `
Object {
  ".gitignore": Object {
    "contents": "node_modules/
dist/
.scp/
.env
Makefile*.mta
mta_archives
mta-*
resources
archive.zip
.*_mta_build_tmp",
    "state": "modified",
  },
  "package.json": Object {
    "contents": "{
  \\"name\\": \\"mta-project\\",
  \\"version\\": \\"0.0.1\\",
  \\"description\\": \\"Build and deployment scripts\\",
  \\"scripts\\": {
    \\"clean\\": \\"rimraf resources mta_archives mta-op*\\",
    \\"build\\": \\"rimraf resources mta_archives && mbt build --mtar archive\\",
    \\"deploy\\": \\"cf deploy mta_archives/archive.mtar --retries 1\\",
    \\"undeploy\\": \\"cf undeploy appfrontend --delete-services --delete-service-keys --delete-service-brokers\\"
  },
  \\"devDependencies\\": {
    \\"mbt\\": \\"^1.2.29\\",
    \\"rimraf\\": \\"^5.0.5\\"
  }
}
",
    "state": "modified",
  },
  "xs-security.json": Object {
    "contents": "{
  \\"xsappname\\": \\"appfrontend\\",
  \\"tenant-mode\\": \\"dedicated\\",
  \\"description\\": \\"Security profile of called application\\",
  \\"scopes\\": [],
  \\"role-templates\\": []
}
",
    "state": "modified",
  },
}
`;

exports[`CF Writer Base Generate Base Config - App Frontend Generate deployment configs - app frontend 2`] = `
"_schema-version: \\"3.2\\"
ID: appfrontend
description: MyManagedDescription
version: 0.0.1
modules:
- name: appfrontend-app-content
  type: com.sap.application.content
  path: .
  requires:
  - name: appfrontend-app-front
    parameters:
      content-target: true
  - name: appfrontend-uaa
  parameters:
    config:
      destinations:
      - name: ui5
        url: https://ui5.sap.com
resources:
- name: appfrontend-uaa
  type: org.cloudfoundry.managed-service
  parameters:
    config:
      tenant-mode: dedicated
      xsappname: appfrontend-\${org}-\${space}
    path: ./xs-security.json
    service: xsuaa
    service-name: appfrontend-xsuaa-service
    service-plan: application
- name: appfrontend-app-front
  type: org.cloudfoundry.managed-service
  parameters:
    service: app-front
    service-name: appfrontend-app-front-service
    service-plan: developer
parameters:
  deploy_mode: html5-repo
  enable-parallel-deployments: true
"
`;

exports[`CF Writer Base Generate Base Config - Managed Generate deployment configs - managed 1`] = `
Object {
  ".gitignore": Object {
    "contents": "node_modules/
dist/
.scp/
.env
Makefile*.mta
mta_archives
mta-*
resources
archive.zip
.*_mta_build_tmp",
    "state": "modified",
  },
  "package.json": Object {
    "contents": "{
  \\"name\\": \\"mta-project\\",
  \\"version\\": \\"0.0.1\\",
  \\"description\\": \\"Build and deployment scripts\\",
  \\"scripts\\": {
    \\"clean\\": \\"rimraf resources mta_archives mta-op*\\",
    \\"build\\": \\"rimraf resources mta_archives && mbt build --mtar archive\\",
    \\"deploy\\": \\"cf deploy mta_archives/archive.mtar --retries 1\\",
    \\"undeploy\\": \\"cf undeploy managed --delete-services --delete-service-keys --delete-service-brokers\\"
  },
  \\"devDependencies\\": {
    \\"mbt\\": \\"^1.2.29\\",
    \\"rimraf\\": \\"^5.0.5\\"
  }
}
",
    "state": "modified",
  },
  "xs-security.json": Object {
    "contents": "{
  \\"xsappname\\": \\"managed\\",
  \\"tenant-mode\\": \\"dedicated\\",
  \\"description\\": \\"Security profile of called application\\",
  \\"scopes\\": [],
  \\"role-templates\\": []
}
",
    "state": "modified",
  },
}
`;

exports[`CF Writer Base Generate Base Config - Managed Generate deployment configs - managed 2`] = `
"_schema-version: \\"3.2\\"
ID: managed
description: MyManagedDescription
version: 0.0.1
modules:
- name: managed-destination-content
  type: com.sap.application.content
  requires:
  - name: managed-destination-service
    parameters:
      content-target: true
  - name: managed-repo-host
    parameters:
      service-key:
        name: managed-repo-host-key
  - name: managed-uaa
    parameters:
      service-key:
        name: managed-uaa-key
  parameters:
    content:
      instance:
        destinations:
        - Name: managed_html_repo_host
          ServiceInstanceName: managed-html5-service
          ServiceKeyName: managed-repo-host-key
          sap.cloud.service: managed
        - Authentication: OAuth2UserTokenExchange
          Name: managed_uaa
          ServiceInstanceName: managed-xsuaa-service
          ServiceKeyName: managed-uaa-key
          sap.cloud.service: managed
        existing_destinations_policy: update
  build-parameters:
    no-source: true
resources:
- name: managed-destination-service
  type: org.cloudfoundry.managed-service
  parameters:
    config:
      HTML5Runtime_enabled: true
      init_data:
        instance:
          destinations:
          - Authentication: NoAuthentication
            Name: ui5
            ProxyType: Internet
            Type: HTTP
            URL: https://ui5.sap.com
          existing_destinations_policy: update
      version: 1.0.0
    service: destination
    service-name: managed-destination-service
    service-plan: lite
- name: managed-uaa
  type: org.cloudfoundry.managed-service
  parameters:
    path: ./xs-security.json
    service: xsuaa
    service-name: managed-xsuaa-service
    service-plan: application
- name: managed-repo-host
  type: org.cloudfoundry.managed-service
  parameters:
    service: html5-apps-repo
    service-name: managed-html5-service
    service-plan: app-host
parameters:
  deploy_mode: html5-repo
  enable-parallel-deployments: true
"
`;

exports[`CF Writer Base Generate Base Config - Standalone Generate deployment configs - standalone with ABAP service provider 1`] = `
Object {
  ".gitignore": Object {
    "contents": "node_modules/
dist/
.scp/
.env
Makefile*.mta
mta_archives
mta-*
resources
archive.zip
.*_mta_build_tmp",
    "state": "modified",
  },
  "package.json": Object {
    "contents": "{
  \\"name\\": \\"mta-project\\",
  \\"version\\": \\"0.0.1\\",
  \\"description\\": \\"Build and deployment scripts\\",
  \\"scripts\\": {
    \\"clean\\": \\"rimraf resources mta_archives mta-op*\\",
    \\"build\\": \\"rimraf resources mta_archives && mbt build --mtar archive\\",
    \\"deploy\\": \\"cf deploy mta_archives/archive.mtar --retries 1\\",
    \\"undeploy\\": \\"cf undeploy standalonewithabapserviceprovider --delete-services --delete-service-keys --delete-service-brokers\\"
  },
  \\"devDependencies\\": {
    \\"mbt\\": \\"^1.2.29\\",
    \\"rimraf\\": \\"^5.0.5\\"
  }
}
",
    "state": "modified",
  },
  "router/package.json": Object {
    "contents": "{
    \\"name\\": \\"app-router\\",
    \\"private\\": true,
    \\"description\\": \\"App router\\",
    \\"engines\\": {
        \\"node\\": \\">= 16.0.0\\"
    },
    \\"scripts\\": {
        \\"start\\": \\"node node_modules/@sap/approuter/approuter.js\\",
        \\"start-local\\": \\"node node_modules/@sap/html5-repo-mock/index.js\\"
    },
    \\"dependencies\\": {
        \\"@sap/approuter\\": \\"^19\\"
    },
    \\"devDependencies\\": {
        \\"@sap/html5-repo-mock\\": \\"^2.1.0\\"
    }
}
",
    "state": "modified",
  },
  "router/xs-app.json": Object {
    "contents": "{
  \\"authenticationMethod\\": \\"route\\",
  \\"routes\\": [
    {
      \\"source\\": \\"^/sap/(.*)$\\",
      \\"target\\": \\"/sap/$1\\",
      \\"service\\": \\"TestService\\",
      \\"endpoint\\": \\"TestEndPoint\\",
      \\"authenticationType\\": \\"xsuaa\\",
      \\"csrfProtection\\": false
    },
    {
      \\"source\\": \\"^/resources/(.*)$\\",
      \\"target\\": \\"/resources/$1\\",
      \\"authenticationType\\": \\"none\\",
      \\"destination\\": \\"ui5\\"
    },
    {
      \\"source\\": \\"^/test-resources/(.*)$\\",
      \\"target\\": \\"/test-resources/$1\\",
      \\"authenticationType\\": \\"none\\",
      \\"destination\\": \\"ui5\\"
    }
  ]
}
",
    "state": "modified",
  },
  "xs-security.json": Object {
    "contents": "{
  \\"xsappname\\": \\"standalonewithabapserviceprovider\\",
  \\"tenant-mode\\": \\"dedicated\\",
  \\"description\\": \\"Security profile of called application\\",
  \\"scopes\\": [],
  \\"role-templates\\": []
}
",
    "state": "modified",
  },
}
`;

exports[`CF Writer Base Generate Base Config - Standalone Generate deployment configs - standalone with ABAP service provider 2`] = `
"_schema-version: \\"3.2\\"
ID: standalonewithabapserviceprovider
description: Generated by Fiori Tools
version: 0.0.1
modules:
- name: standalonewithabapserviceprovider-router
  type: approuter.nodejs
  path: router
  requires:
  - name: standalonewithabapserviceprovider-html5-repo-runtime
  - name: standalonewithabapserviceprovider-uaa
  - name: standalonewithabapserviceprovider-destination-service
    group: destinations
    properties:
      forwardAuthToken: false
      name: ui5
      url: https://ui5.sap.com
  - name: standalonewithabapservic-abap-Y11_00.0035
  parameters:
    disk-quota: 256M
    memory: 256M
resources:
- name: standalonewithabapserviceprovider-uaa
  type: org.cloudfoundry.managed-service
  parameters:
    config:
      tenant-mode: dedicated
      xsappname: standalonewithabapserviceprovider-\${space-guid}
    service: xsuaa
    service-plan: application
- name: standalonewithabapserviceprovider-html5-repo-runtime
  type: org.cloudfoundry.managed-service
  parameters:
    service: html5-apps-repo
    service-plan: app-runtime
- name: standalonewithabapserviceprovider-destination-service
  type: org.cloudfoundry.managed-service
  parameters:
    config:
      HTML5Runtime_enabled: false
      init_data:
        instance:
          destinations:
          - Authentication: NoAuthentication
            Name: ui5
            ProxyType: Internet
            Type: HTTP
            URL: https://ui5.sap.com
          existing_destinations_policy: update
      version: 1.0.0
    service: destination
    service-name: standalonewithabapserviceprovider-destination-service
    service-plan: lite
- name: standalonewithabapservic-abap-Y11_00.0035
  type: org.cloudfoundry.existing-service
  parameters:
    protocol:
    - ODataV2
    service: abap-haas
    service-name: Y11_00.0035
    service-plan: 16_abap_64_db
parameters:
  deploy_mode: html5-repo
  enable-parallel-deployments: true
"
`;

exports[`CF Writer Base Generate Base Config - Standalone Generate deployment configs - standalone with connectivity service 1`] = `
Object {
  ".gitignore": Object {
    "contents": "node_modules/
dist/
.scp/
.env
Makefile*.mta
mta_archives
mta-*
resources
archive.zip
.*_mta_build_tmp",
    "state": "modified",
  },
  "package.json": Object {
    "contents": "{
  \\"name\\": \\"mta-project\\",
  \\"version\\": \\"0.0.1\\",
  \\"description\\": \\"Build and deployment scripts\\",
  \\"scripts\\": {
    \\"clean\\": \\"rimraf resources mta_archives mta-op*\\",
    \\"build\\": \\"rimraf resources mta_archives && mbt build --mtar archive\\",
    \\"deploy\\": \\"cf deploy mta_archives/archive.mtar --retries 1\\",
    \\"undeploy\\": \\"cf undeploy standalone-with-connectivity-service --delete-services --delete-service-keys --delete-service-brokers\\"
  },
  \\"devDependencies\\": {
    \\"mbt\\": \\"^1.2.29\\",
    \\"rimraf\\": \\"^5.0.5\\"
  }
}
",
    "state": "modified",
  },
  "router/package.json": Object {
    "contents": "{
    \\"name\\": \\"app-router\\",
    \\"private\\": true,
    \\"description\\": \\"App router\\",
    \\"engines\\": {
        \\"node\\": \\">= 16.0.0\\"
    },
    \\"scripts\\": {
        \\"start\\": \\"node node_modules/@sap/approuter/approuter.js\\",
        \\"start-local\\": \\"node node_modules/@sap/html5-repo-mock/index.js\\"
    },
    \\"dependencies\\": {
        \\"@sap/approuter\\": \\"^19\\"
    },
    \\"devDependencies\\": {
        \\"@sap/html5-repo-mock\\": \\"^2.1.0\\"
    }
}
",
    "state": "modified",
  },
  "router/xs-app.json": Object {
    "contents": "{
  \\"authenticationMethod\\": \\"route\\",
  \\"routes\\": [
    {
      \\"source\\": \\"^(?:/app|/app/.*)?/resources/(.*)$\\",
      \\"target\\": \\"/resources/$1\\",
      \\"authenticationType\\": \\"none\\",
      \\"destination\\": \\"ui5\\"
    },
    {
      \\"source\\": \\"^(?:/app|/app/.*)?/test-resources/(.*)$\\",
      \\"target\\": \\"/test-resources/$1\\",
      \\"authenticationType\\": \\"none\\",
      \\"destination\\": \\"ui5\\"
    }
  ]
}
",
    "state": "modified",
  },
  "xs-security.json": Object {
    "contents": "{
  \\"xsappname\\": \\"standalone-with-connectivity-service\\",
  \\"tenant-mode\\": \\"dedicated\\",
  \\"description\\": \\"Security profile of called application\\",
  \\"scopes\\": [],
  \\"role-templates\\": []
}
",
    "state": "modified",
  },
}
`;

exports[`CF Writer Base Generate Base Config - Standalone Generate deployment configs - standalone with connectivity service 2`] = `
"_schema-version: \\"3.2\\"
ID: standalone-with-connectivity-service
description: Generated by Fiori Tools
version: 0.0.1
modules:
- name: standalone-with-connectivity-service-router
  type: approuter.nodejs
  path: router
  requires:
  - name: standalone-with-connectivity-service-html5-repo-runtime
  - name: standalone-with-connectivity-service-uaa
  - name: standalone-with-connectivity-service-destination-service
    group: destinations
    properties:
      forwardAuthToken: false
      name: ui5
      url: https://ui5.sap.com
  - name: standalone-with-connectivity-service-connectivity
  parameters:
    disk-quota: 256M
    memory: 256M
resources:
- name: standalone-with-connectivity-service-uaa
  type: org.cloudfoundry.managed-service
  parameters:
    config:
      tenant-mode: dedicated
      xsappname: standalone-with-connectivity-service-\${space-guid}
    service: xsuaa
    service-plan: application
- name: standalone-with-connectivity-service-html5-repo-runtime
  type: org.cloudfoundry.managed-service
  parameters:
    service: html5-apps-repo
    service-plan: app-runtime
- name: standalone-with-connectivity-service-destination-service
  type: org.cloudfoundry.managed-service
  parameters:
    config:
      HTML5Runtime_enabled: false
      init_data:
        instance:
          destinations:
          - Authentication: NoAuthentication
            Name: ui5
            ProxyType: Internet
            Type: HTTP
            URL: https://ui5.sap.com
          existing_destinations_policy: update
      version: 1.0.0
    service: destination
    service-name: standalone-with-connectivity-service-destination-service
    service-plan: lite
- name: standalone-with-connectivity-service-connectivity
  type: org.cloudfoundry.managed-service
  parameters:
    service: connectivity
    service-plan: lite
parameters:
  deploy_mode: html5-repo
  enable-parallel-deployments: true
"
`;
